gax-grpc/src/main/java/com/google/api/gax/grpc/OperationFuture.java

```diff
@@ -27,116 +27,30 @@
- * the {@link com.google.api.gax.retrying.TimedRetryAlgorithm} specified in {@link
- * OperationCallSettings} provided during creation of the corresponding {@link OperationCallable}.
- *
- * <p>This class is thread-safe.
- */
+/** An ApiFuture which provides access to additional operation data. */
 @BetaApi
-public final class OperationFuture<ResponseT extends Message, MetadataT extends Message>
-    implements ApiFuture<ResponseT> {
-  private final Object lock = new Object();
-
-  private final RetryingFuture<Operation> pollingFuture;
-  private final ApiFuture<Operation> initialFuture;
-  private final ApiFuture<ResponseT> resultFuture;
-  private final ApiFunction<Operation, MetadataT> metadataTransformer;
-
-  private volatile ApiFuture<Operation> peekedAttemptResult;
-  private volatile ApiFuture<MetadataT> peekedPollResult;
-  private volatile ApiFuture<Operation> gottenAttemptResult;
-  private volatile ApiFuture<MetadataT> gottenPollResult;
-
-  /**
-   * Creates a new operation future instance.
-   *
-   * @param pollingFuture retrying future which tracks polling of the server operation (in most
-   *     cases with exponential upper bounded intervals)
-   * @param initialFuture the initial future which started the operation on the server side
-   * @param responseTransformer transformer which unpacks a {@link ResponseT} object from an {@link
-   *     Operation} object
-   * @param metadataTransformer transformer which unpacks a {@link MetadataT} object from an {@link
-   *     Operation} object
-   */
-  public OperationFuture(
-      RetryingFuture<Operation> pollingFuture,
-      ApiFuture<Operation> initialFuture,
-      ApiFunction<Operation, ResponseT> responseTransformer,
-      ApiFunction<Operation, MetadataT> metadataTransformer) {
-    this.pollingFuture = checkNotNull(pollingFuture);
-    this.initialFuture = checkNotNull(initialFuture);
-    this.resultFuture = ApiFutures.transform(pollingFuture, responseTransformer);
-    this.metadataTransformer = checkNotNull(metadataTransformer);
-  }
-
-  @Override
-  public void addListener(Runnable listener, Executor executor) {
-    pollingFuture.addListener(listener, executor);
-  }
-
-  @Override
-  public boolean cancel(boolean mayInterruptIfRunning) {
-    return pollingFuture.cancel(mayInterruptIfRunning);
-  }
-
-  @Override
-  public boolean isCancelled() {
-    return pollingFuture.isCancelled();
-  }
-
-  @Override
-  public boolean isDone() {
-    return pollingFuture.isDone();
-  }
-
-  @Override
-  public ResponseT get() throws InterruptedException, ExecutionException {
-    pollingFuture.get();
-    return resultFuture.get();
-  }
-
-  @Override
-  public ResponseT get(long timeout, TimeUnit unit)
-      throws InterruptedException, ExecutionException, TimeoutException {
-    pollingFuture.get(timeout, unit);
-    return resultFuture.get();
-  }
-
+public abstract class OperationFuture<ResponseT, MetadataT, OperationT>
```

Comment:
There are also many disadvantages. 
Also in a long term we will be able to use default methods (java 8).
This is actually first time in my life I see abstract class with abstract -only methods. 
And at least, please make it implement ApiFuture, instead of extending AbstractApiFuture. AbstractApiFuture adds unndecessary layering. We should care about layering here: it was a complete nightmare to organize the callbacks for retryingfuture/operationfuture. 

The callback ordering, looping of callbacks, handling/propagating exceptions from callbacks/caused by callbacks etc. And all that in heavily multithreaded env.

Another **big issure** here is the fact that AbstractApiFuture has set() and setException() methods **public** here. It is as dangerous here as the client now can get OperationFuture and set its result itself from the outside. This is fixed in newer version of api-common, but we haven't released that yet.